rfd/0152-automatic-database-users-mongodb.md

```diff
@@ -0,0 +1,202 @@
+in-database connections. Instead, one can authenticate with Atlas using the
+Atlas SDK and use APIs to manage these database users and roles. Multiple
+deployment jobs will be created to update the MongoDB clusters in this project,
+upon successful APIs calls.
+
+In my personal testing on an Atlas project with a single MongoDB cluster, it
+takes 10~20 seconds for the deployment job to refresh the database user in the
+target MongoDB cluster.
+
+With the current design of Automatic User Provisioning, the database user must
+be updated with new role assignments for each new connection then the roles
+should be revoked once the connection is done. However, waiting for 10+ seconds
+for provisioning the database user each connection will result in a very bad
+user experience (also client may just time out).
+
+## Self-hosted MongoDB Details
+
+The overall flow and logic will follow the previous [RFD
+113](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md).
+Differences will be outlined in the sections below.
+
+### The admin user connection
+
+The Database Service will connect as an admin user in order to manage database
+users. The admin user requires the following privileges:
+```json
+{
+  "createUser": "CN=teleport-admin",
+  "roles": [
+    { "role": "clusterMonitor", "db": "admin" },
+    { "role": "userAdminAnyDatabase", "db": "admin" }
+  ]
+}
+```
+
+As the implementation of other databases, the name of admin user is
+defined in the database spec `admin_user.name`, and the admin user will
+authenticate using X.509 for self-hosted databases.
+
+However, there is NO concept of "Stored Procedures" in MongoDB. The user
+provisioning logic will be carried through multiple `runCommand` calls
+implemented in Go. Multiple parallel database sessions will not race thanks to
+[semaphore
+locking](https://github.com/gravitational/teleport/blob/master/rfd/0113-automatic-database-users.md#locking).
+
+### Roles
+
+Unlike MySQL or PostgreSQL where roles are scoped to the entire database
+instance/cluster, a role in MongoDB is scoped to a specific database.
+
+For example, a custom role `myCustomRole` can be created on database `db1` with
+specific privileges on `db1`, and another role `myCustomRole` can be created on
+database `db2` with specific privileges on `db2`. They will be considered two
+different roles.
+
+Most built-in roles are available on all databases, while some built-in roles
+like `readAnyDatabase` can only be applied on the `admin` database.
+
+Therefore, when specifying database roles to assign for the user, it must be in
+the format of `<role-name>@<db-name>` to fully identify the role:
+
+```yaml
+kind: "role"
+version: "v6"
+metadata:
+  name: "example"
+spec:
+  options:
+    create_db_user_mode: keep
+  allow:
+    db_names:
+    - "db1"
+    - "db2"
+    - "db3"
+    db_roles: 
+    - "readAnyDatabase@admin"
+    - "readWrite@db2"
+    - "myCustomRole@db3"
```

Comment:
Is there any whey to support `db_names:  "*"` schema where database are not known ?

Reply:
`db_names` can be "*" as this is what Teleport restricts the user. But `db_roles` must be in role+db combo, as this is what Mongo Server restricts the user.

I did about whether we can have a `read` or `read@*` as `db_roles`. but we have to fill the database somehow as we must specify the role+db combo when creating/updating the user. we can fill the database from `db_names` or maybe take the `db_name` from db route. but both methods have their own issues. So i think it's clearer to always do `role@db` as `db_roles`.